subsys/audio_module/audio_module.c

```diff
@@ -0,0 +1,1250 @@
+		ret = data_fifo_pointer_first_vacant_get(rx_handle->thread.msg_rx,
+							 (void **)&data_msg_rx, K_NO_WAIT);
+		if (ret) {
+			LOG_ERR("Module %s no free data buffer, ret %d", rx_handle->name, ret);
+			return ret;
+		}
+
+		/* Copy. The audio data itself will remain in its original location. */
+		memcpy(&data_msg_rx->audio_data, audio_data, sizeof(struct audio_data));
+		data_msg_rx->tx_handle = tx_handle;
+		data_msg_rx->response_cb = data_in_response_cb;
+
+		ret = data_fifo_block_lock(rx_handle->thread.msg_rx, (void **)&data_msg_rx,
+					   sizeof(struct audio_module_message));
+		if (ret) {
+			data_fifo_block_free(rx_handle->thread.msg_rx, (void **)&data_msg_rx);
+
+			LOG_WRN("Module %s failed to queue audio data, ret %d", rx_handle->name,
+				ret);
+			return ret;
+		}
+
+		LOG_DBG("Audio data sent to module %s", rx_handle->name);
+
+	} else {
+		LOG_WRN("Receiving module %s is in an invalid state %d", rx_handle->name,
+			rx_handle->state);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Send audio data item to the module's TX FIFO.
+ *
+ * @param handle[in/out]  The handle for this modules instance.
+ * @param audio_data[in]  A pointer to the audio data.
+ *
+ * @return 0 if successful, error value.
+ */
+static int tx_fifo_put(struct audio_module_handle *handle,
+		       struct audio_data const *const audio_data)
+{
+	int ret;
+	struct audio_module_message *data_msg_tx;
+
+	/* Take a TX message. */
+	ret = data_fifo_pointer_first_vacant_get(handle->thread.msg_tx, (void **)&data_msg_tx,
+						 K_NO_WAIT);
+	if (ret) {
+		LOG_WRN("No free space in TX FIFO for module %s, ret %d", handle->name, ret);
+		return ret;
+	}
+
+	/* Configure audio data. */
+	memcpy(&data_msg_tx->audio_data, audio_data, sizeof(struct audio_data));
+	data_msg_tx->tx_handle = handle;
+	data_msg_tx->response_cb = audio_data_release_cb;
+
+	/* Send audio data to modules output message queue. */
+	ret = data_fifo_block_lock(handle->thread.msg_tx, (void **)&data_msg_tx,
+				   sizeof(struct audio_module_message));
+	if (ret) {
+		LOG_ERR("Failed to send audio data to output of module %s, ret %d", handle->name,
+			ret);
+
+		data_fifo_block_free(handle->thread.msg_tx, (void **)&data_msg_tx);
+
+		ret = k_sem_take(&handle->sem, K_NO_WAIT);
+		if (ret) {
+			LOG_ERR("Failed to take semaphore for data release callback function");
+		}
+
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Send the audio data item to all connected modules.
+ *
+ * @param handle[in/out]  The handle for this modules instance.
+ * @param audio_data[in]  A pointer to the audio data.
+ *
+ * @return 0 if successful, error value.
+ */
+static int send_to_connected_modules(struct audio_module_handle *handle,
+				     struct audio_data const *const audio_data)
+{
+	int ret;
+	struct audio_module_handle *handle_to;
+
+	if (handle->dest_count == 0) {
+		LOG_WRN("Nowhere to send the audio data from module %s so releasing it",
```

Comment:
This should be a debug print?

Reply:
Can we discuss?

Reply:
Hmm, I think I take this one back. Probably best to be WRN?